graph.py

```diff
@@ -0,0 +1,269 @@
+
+    def addVertex(self, continent, country, city, latitude, longitude):
+        self.numVertices = self.numVertices + 1
+        newVertex = Vertex(continent, country, city, latitude, longitude)
+        self.vertList[city] = newVertex
+        return newVertex
+
+    def getVertex(self, n):
+        if n in self.vertList:
+            return self.vertList[n]
+        else:
+            return None
+
+    def edgeExists(self, v1, v2):
+        if v1 in self.getVertex(v2).connectedTo.keys():
+            return True
+        else:
+            return False
+
+    def getDistance(self, city1, city2):
+
+        lat1 = self.vertList[city1].latitude
+        long1 = self.vertList[city1].longitude
+
+        lat2 = self.vertList[city2].latitude
+        long2 = self.vertList[city2].longitude
+
+        lat_dist = lat2 - lat1
+        long_dist = long2 - long1
+
+        return math.hypot(lat_dist, long_dist)
+
+    def __contains__(self,n):
+        return n in self.vertList
+
+    def addEdge(self, f, t, weight=0):
+        if f not in self.vertList:
+            nv = self.addVertex(f)
+        if t not in self.vertList:
+            nv = self.addVertex(t)
+        self.vertList[f].addNeighbor(self.vertList[t], weight)
+
+    def getVertices(self):
+        return self.vertList.keys()
+
+    def __iter__(self):
+        return iter(self.vertList.values())
+
+    def dijkstra(self, src, dest):
+
+        assert src in self.vertList, 'Such source node doesn\'t exist'
+
+        distances = {vertex: np.Infinity for vertex in self.vertList}
+        distances[src] = 0
+
+        previous_vertices = {
+            vertex: None for vertex in self.vertList
+        }
+
+        vertices = self.vertList.copy()
+
+        while vertices:
+
+            current_vertex = min(
+                vertices, key=lambda vertex: distances[vertex])
+
+            if distances[current_vertex] == np.Infinity:
+                break
+
+            # Find unvisited neighbors for the current node
+            # and calculate their distances through the current node.
+            for neighbor in self.getVertex(current_vertex).connectedTo:
+
+                cost = self.getVertex(current_vertex).connectedTo[neighbor]
+                possible_path = distances[current_vertex] + cost
+
+                if possible_path < distances[neighbor.id]:
+
+                    distances[neighbor.id] = possible_path
+                    previous_vertices[neighbor.id] = current_vertex
+
+            del(vertices[current_vertex])
+
+        path, current_vertex = deque(), dest
+
+        while previous_vertices[current_vertex] is not None:
+            path.appendleft(current_vertex)
+            current_vertex = previous_vertices[current_vertex]
+        if path:
+            path.appendleft(current_vertex)
+        return path
+
+
+    def two_opt(self, tour):
+        pass
+
+    def is_path_complete(self, path):
+        pass
+
+
+    def greedyTSP(self):
```

Comment:
So this is the solution and so if you keep it here, it'll be part of the workshop which kinda defeats the purpose of the exercise.

What I would recommend is splitting the code in two: a library that contains the basic graph primitives used to implement a solution, and check the solution for correctness plus debugging utilities to make everyone's life easier. The second should be a simple brute force executable that shows how to use the library and giving a starting point to improve the solution.

I also recommend adding a read me with basic information for reference like what's TSP and it's general properties.

Reply:
Note that you keep your solution in a branch and maybe have people working on the workshop submit their solution in PRs to your repo. You could then even build a test harness to compare the various solutions together and declare a winner at the end.